baseapp/baseapp.go

```diff
@@ -683,7 +685,12 @@ func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*s
 		}
 
 		msgRoute := msg.Route()
-		handler := app.router.Route(ctx, msgRoute)
+		// First, check if the MsgService router handles the route.
+		handler := app.MsgServiceRouter().Route(msgRoute)
+		// If not, then check if the legacy router handles the route.
+		if handler == nil {
+			handler = app.router.Route(ctx, msgRoute)
+		}
```

Comment:
regarding https://github.com/cosmos/cosmos-sdk/pull/7519#discussion_r503415434, I was thinking something like this.

Reply:
This should either be dealt with outside of `BaseApp` (see `Routing in ADR 031)` or there should be a type assertion `msg.(ServiceMsg)` and a whole separate if branch here rather than using `Route`.